File: shared/java/top/fpsmaster/modules/music/JLayerHelper.java
package top.fpsmaster.modules.music;

import edu.emory.mathcs.jtransforms.fft.DoubleFFT_1D;
import javazoom.jl.converter.Converter;
import javazoom.jl.decoder.JavaLayerException;

import javax.sound.sampled.*;
import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import static java.lang.Math.min;
import static java.lang.Math.sqrt;

public class JLayerHelper {
    public static Clip clip;
    private static AudioInputStream audIn;
    private static byte[] audioBytes = new byte[0];
    public static double[] loudnessCurve = new double[0];

    public static float getProgress() {
        if (clip == null) return 0;
        long timeElapsed = clip.getMicrosecondPosition();
        long total = clip.getMicrosecondLength();
        return (float) timeElapsed / total;
    }

    public static void playWAV(String wavFile) throws IOException, LineUnavailableException {
        File soundFile = new File(wavFile);
        try {
            AudioInputStream aud = AudioSystem.getAudioInputStream(soundFile);
            audIn = AudioSystem.getAudioInputStream(soundFile);
            audioBytes = readAudioData(audIn);
            clip = AudioSystem.getClip();
            if (clip == null) return;
            clip.open(aud);
            clip.start();
        } catch (UnsupportedAudioFileException e) {
            e.printStackTrace();
        }
    }

    public static void seek(float progress) {
        long totalTime = clip.getMicrosecondLength();
        long currentTime = (long) (totalTime * progress);
        clip.setMicrosecondPosition(currentTime);
    }

    public static void updateLoudness() {
        if (clip == null || audIn == null) return;

        AudioFormat format = audIn.getFormat();
        if (format.getEncoding() != AudioFormat.Encoding.PCM_SIGNED || format.getSampleSizeInBits() != 16) {
            return;
        }

        double currentTimeSec = clip.getMicrosecondPosition() / 1_000_000.0;
        int fftSize = 1024;
        double fftWindowDuration = fftSize / format.getSampleRate();

        byte[] audioSegment = getAudioSegment(
                audioBytes,
                format.getSampleRate(),
                format.getFrameSize(),
                (float) (currentTimeSec - fftWindowDuration / 2),
                (float) fftWindowDuration
        );

        double[] fftData = performFFT(audioSegment);
        double[] amplitudes = computeAmplitude(fftData);

        loudnessCurve = amplitudes;
    }

    private static byte[] readAudioData(AudioInputStream audioInputStream) throws IOException {
        int bufferSize = (int) (audioInputStream.getFrameLength() * audioInputStream.getFormat().getFrameSize());
        byte[] audioBytes = new byte[bufferSize];
        audioInputStream.read(audioBytes);
        return audioBytes;
    }

    private static byte[] getAudioSegment(byte[] audioData, float sampleRate, int bytesPerFrame, float startSecond, float durationInSeconds) {
        int startSample = (int) (startSecond * sampleRate);
        int numSamples = (int) (durationInSeconds * sampleRate);

        int startByte = startSample * bytesPerFrame;
        int numBytes = numSamples * bytesPerFrame;

        int endByte = Math.min(startByte + numBytes, audioData.length);

        return Arrays.copyOfRange(audioData, startByte, endByte);
    }

    private static double[] performFFT(byte[] buffer) {
        int numSamples = buffer.length / 2;
        double[] audioData = new double[numSamples];

        for (int i = 0; i < numSamples; i++) {
            int low = buffer[i * 2] & 0xff;
            int high = buffer[i * 2 + 1];
            int sample = (high << 8) | low;
            audioData[i] = sample / 32768.0; // Normalize to [-1, 1]
        }

        int fftSize = 1024;
        double[] paddedData = new double[fftSize];
        for (int i = 0; i < min(fftSize, audioData.length); i++) {
            paddedData[i] = audioData[i];
        }

        DoubleFFT_1D fft = new DoubleFFT_1D(fftSize);
        fft.realForward(paddedData);

        return paddedData;
    }

    public static double[] computeAmplitude(double[] fftData) {
        int n = fftData.length;
        double[] amplitudes = new double[n / 2];

        for (int i = 0; i < amplitudes.length; i++) {
            double real = fftData[2 * i];
            double imag = (2 * i + 1 < fftData.length) ? fftData[2 * i + 1] : 0.0;
            amplitudes[i] = sqrt(real * real + imag * imag);
        }

        // Simple normalization
        double maxAmp = Arrays.stream(amplitudes).max().orElse(1.0);
        for (int i = 0; i < amplitudes.length; i++) {
            amplitudes[i] /= maxAmp;
        }

        return amplitudes;
    }

    public static void setVolume(float vol) {
        vol /= 2;
        vol += 0.5f;
        FloatControl volumeControl = (FloatControl) clip.getControl(FloatControl.Type.MASTER_GAIN);
        float volume = (volumeControl.getMaximum() - volumeControl.getMinimum()) * vol + volumeControl.getMinimum();
        volumeControl.setValue(volume);
    }

    public static void convert(String sourcePath, String targetPath) {
        try {
            Converter converter = new Converter();
            File sourceFile = new File(sourcePath);
            File targetFile = new File(targetPath);
            converter.convert(sourceFile.getPath(), targetFile.getPath());
        } catch (JavaLayerException e) {
            e.printStackTrace();
        }
    }

    public static void stop() {
        if (clip != null) {
            clip.stop();
        }
    }

    public static void start() {
        if (clip != null) {
            clip.start();
        }
    }

    public static double getDuration() {
        return clip.getMicrosecondLength() / 1000000.0 / 60.0;
    }
}

File: shared/java/top/fpsmaster/modules/music/JLayerHelper.kt
package top.fpsmaster.modules.music

import edu.emory.mathcs.jtransforms.fft.DoubleFFT_1D
import javazoom.jl.converter.Converter
import javazoom.jl.decoder.JavaLayerException
import java.io.File
import java.io.IOException
import javax.sound.sampled.*
import kotlin.math.min
import kotlin.math.sqrt

object JLayerHelper {
    var clip: Clip? = null
    var audIn: AudioInputStream? = null
    var audioBytes: ByteArray = ByteArray(0)
    var loudnessCurve: DoubleArray? = DoubleArray(0)
    val progress: Float
        get() {
            val timeElapsed = clip!!.microsecondPosition
            val total = clip!!.microsecondLength
            return timeElapsed.toFloat() / total
        }

    fun playWAV(wavFile: String) {
        // Open an audio input stream.
        val soundFile = File(wavFile) //you could also get the sound file with an URL
        var audioIn = AudioSystem.getAudioInputStream(soundFile)
        audIn = AudioSystem.getAudioInputStream(soundFile)
        audioBytes = readAudioData(audIn!!)
        // Get a sound clip resource.
        clip = AudioSystem.getClip()
        // Open audio clip and load samples from the audio input stream.
        if (clip == null)
            return
        clip!!.open(audioIn)
        clip!!.start()
    }

    @JvmStatic
    fun seek(progress: Float) {
        val totalTime = clip!!.microsecondLength
        val currentTime = (totalTime * progress).toLong() //将播放进度设置为50%
        clip!!.microsecondPosition = currentTime
    }

    fun updateLoudness() {
        if (clip == null || audIn == null) return

        val format = audIn!!.format
        if (format.encoding != AudioFormat.Encoding.PCM_SIGNED || format.sampleSizeInBits != 16) {
            println("Unsupported format: $format")
            return
        }

        val currentTimeSec = clip!!.microsecondPosition / 1_000_000.0
        val fftSize = 1024
        val fftWindowDuration = fftSize / format.sampleRate

        val audioSegment = getAudioSegment(
            audioBytes,
            format.sampleRate,
            format.frameSize,
            (currentTimeSec - fftWindowDuration / 2).toFloat(),
            fftWindowDuration.toFloat()
        )

        val fftData = performFFT(audioSegment)
        val amplitudes = computeAmplitude(fftData)

        loudnessCurve = amplitudes
    }


    @Throws(IOException::class)
    fun readAudioData(audioInputStream: AudioInputStream): ByteArray {
        val bufferSize = (audioInputStream.frameLength * audioInputStream.format.frameSize).toInt()
        val audioBytes = ByteArray(bufferSize)
        audioInputStream.read(audioBytes)
        return audioBytes
    }

    fun getAudioSegment(
        audioData: ByteArray,
        sampleRate: Float,
        bytesPerFrame: Int,
        startSecond: Float,
        durationInSeconds: Float
    ): ByteArray {
        val startSample = (startSecond * sampleRate).toInt().coerceAtLeast(0)
        val numSamples = (durationInSeconds * sampleRate).toInt()

        val startByte = startSample * bytesPerFrame
        val numBytes = numSamples * bytesPerFrame

        val endByte = (startByte + numBytes).coerceAtMost(audioData.size)

        return audioData.copyOfRange(startByte, endByte)
    }



    private fun performFFT(buffer: ByteArray): DoubleArray {
        val numSamples = buffer.size / 2
        val audioData = DoubleArray(numSamples)

        // Convert byte pairs (little-endian) to signed 16-bit samples
        for (i in 0 until numSamples) {
            val low = buffer[i * 2].toInt() and 0xff
            val high = buffer[i * 2 + 1].toInt()
            val sample = (high shl 8) or low
            audioData[i] = sample / 32768.0 // Normalize to [-1, 1]
        }

        // Zero-padding to nearest power of 2 (optional, or cut to fixed size like 1024)
        val fftSize = 1024
        val paddedData = DoubleArray(fftSize)
        for (i in 0 until min(fftSize, audioData.size)) {
            paddedData[i] = audioData[i]
        }

        val fft = DoubleFFT_1D(fftSize)
        fft.realForward(paddedData)

        return paddedData
    }



    fun computeAmplitude(fftData: DoubleArray): DoubleArray {
        val n = fftData.size
        val amplitudes = DoubleArray(n / 2)

        for (i in amplitudes.indices) {
            val real = fftData[2 * i]
            val imag = if (2 * i + 1 < fftData.size) fftData[2 * i + 1] else 0.0
            amplitudes[i] = sqrt(real * real + imag * imag)
        }

        // 简单归一化
        val maxAmp = amplitudes.maxOrNull() ?: 1.0
        return amplitudes.map { it / maxAmp }.toDoubleArray()
    }


    fun setVolume(vol: Float) {
        var vol = vol
        vol /= 2
        vol += 0.5f
        val volumeControl = clip!!.getControl(FloatControl.Type.MASTER_GAIN) as FloatControl
        // Change the volume to half way between minimum and maximum
        val volume = (volumeControl.maximum - volumeControl.minimum) * vol + volumeControl.minimum
        volumeControl.value = volume
    }

    fun convert(sourcePath: String, targetPath: String) {
        try {
            val converter = Converter()
            val sourceFile = File(sourcePath)
            val targetFile = File(targetPath)
            converter.convert(sourceFile.path, targetFile.path)
        } catch (e: JavaLayerException) {
            e.printStackTrace()
        }
    }

    fun stop() {
        clip!!.stop()
    }

    fun start() {
        clip!!.start()
    }

    @JvmStatic
    val duration: Double
        get() = (clip!!.microsecondLength / 1000f / 1000f / 60f).toDouble()
}
